src/media/media.ts

```diff
@@ -0,0 +1,1041 @@
+   *      for invalid resolution requests. If `video` was requested with a
+   *      `videoResolution` value (could be a SDK default), it will retry
+   *      video with the same passed in value but with _no_ resolution.
+   *  3. If a deviceId was requested and there was a failure, this will retry
+   *      media with the SDK default deviceId for that media type.
+   *  4. If the SDK default deviceId fails (or it didn't exist), then this
+   *      will retry with system defaults and no other options (such as `deviceId`s,
+   *      `videoFrameRate`, & `videoResolution`)
+   *  5. If system defaults fails, it will throw the error and stop attempting
+   *      to retry.
+   *
+   * Note: if using `retryOnFailure` it is recommended to get the track labels of
+   *  the returned tracks to ensure you received the desired device. This works in
+   *  line with the browser option of `deviceId: { ideal: 'device-id-string' }`.
+   *
+   * Warning: if `sdk.media.requestPermissions(type)` has NOT been called before
+   *  calling `startMedia`, `startMedia` will call `sdk.media.requestPermissions(type)`.
+   *  If calling `startMedia` with both `audio` and `video` _before_ requesting permissions,
+   *  `startMedia` will attempt to gain permissions for `audio` first and then `video` (because
+   *  media permissions must be requested one at a time). If `audio` fails, it will
+   *  not attempt to gain permissions for `video` – the error will stop execution.
+   *
+   * @param mediaReqOptions video and/or audio default device or deviceId. Defaults to `{video: true, audio: true}`
+   * @param retryOnFailure whether the sdk should retry on an error
+   */
+  async startMedia (mediaReqOptions: IMediaRequestOptions = { video: true, audio: true }, retryOnFailure: boolean = true): Promise<MediaStream> {
+    /* `getStandardConstraints` will set media type to `truthy` if `null` was passed in */
+    const requestingVideo = mediaReqOptions.video || mediaReqOptions.video === null;
+    const requestingAudio = mediaReqOptions.audio || mediaReqOptions.audio === null;
+    const conversationId = mediaReqOptions.session?.conversationId;
+    const sessionId = mediaReqOptions.session?.id;
+    const {
+      micPermissionsRequested,
+      cameraPermissionsRequested
+    } = this.getState();
+
+    const loggingExtras = {
+      mediaReqOptions: { ...mediaReqOptions, session: undefined },
+      retryOnFailure,
+      conversationId,
+      sessionId,
+      micPermissionsRequested,
+      cameraPermissionsRequested
+    };
+
+    let mediaStream: MediaStream;
+
+    this.sdk.logger.info('calling sdk.media.startMedia()', loggingExtras);
+
+    /* if we aren't requesting any media, call through to gUM to throw an error */
+    if (!requestingAudio && !requestingVideo) {
+      /* 'none' will throw the error we want so make sure to set `retry` to `false` */
+      return this.startSingleMedia('none', mediaReqOptions, false);
```

Comment:
why do this instead of just throwing?

Reply:
I thought about create an error and throwing it. But I decided to go this route to skip ahead and let gUM throw the error it normally would. Something I was trying to avoid in all of these media shims is mucking with the errors from the browser or trying to create our own errors for different things. I am open to discussion, though. I know this looks pretty hacky. I can pull it out and just throw a normal SdkError

Reply:
After looking at this more, I think it makes sense to throw an SdkError rather than calling through to gUM and letting gUM throw the error. Since the caller is calling an sdk method it means the caller is breaking the sdk contract for the function params. It hasn't got to gUM yet so there is no browser error to muck with. It makes sense to me now to just throw our out error. I'll make that change.